Use plugin-content-docs instead of duplicating old code + fixes and features

Cyriuz opened the pull request:
This fixes a lot of issues like auto generated sidebars and support for all new sidebar features like doc links on categories, as well as sidebar info specified in the md files header. It also simplifies version handling.

Separated document titles from document ids as the id might not always be the same as the title. And it also fixes the download button not finding the files if the title/id/url doesn't match as is possible with the slug option on sidebar entries.

Fixed toc page numbering for unnumbered headings.

Fixed so that it works for exclusive docs setups with routeBasePath: /

Added options for:
- versions, to replace the functionality to specify which versions to generate for from rootDocIds.
- useExtraPaths to use with the express server.
- alwaysIncludeSiteStyles to be able to add both custom styles and the docusaurus ones.
- ignoreCssSelectors to be able to remove html elements before generating the pdf.
- margins and coverMargins as they we're hardcoded.
- jQueryUrl option in case you want to provide it from somewhere else.
- getPdfFileName function override to customize the filename of the generated pdfs.

Removed rootDocIds as it can be deduced from the docs plugin data.

A drawback to this new approach is that we depend on data passed to postBuild, so we can't support the custom command anymore. I added a haxx env var BUILD_PDF to trigger the build if autoBuildPdfs is false.

This is marked WIP as I probably broke a lot of stuff, and more a PR just to let you know in case you're interested in it. I'm willing to put some polish into it probably in case you think this is a good idea but would need someone else to try it with a different setup.

Cyriuz commented on src/index.ts:
```diff
@@ -39,8 +39,6 @@ export default function (
 
       return {
         headTags: [`
-        <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
```
Forgot I removed this as I already had it in the wrapping site, it should be an option I guess.

Cyriuz commented on src/index.ts:
```diff
@@ -56,22 +54,49 @@ export default function (
-              pdfPath = pdfPath.substr(0, lastSlashPos);
+            var id = getActiveDocumentId();
+
+            while (!pdfPath.endsWith(id) && id != "" && id.includes("/")) {
```
All of this is sketchy at best, this should probably be a theme plugin so we have access to the actual ids.

simologos commented:
Thank you @Cyriuz for working on this! Let me check with some other users of the plugin and come back.

Cyriuz commented:
Took some extra time to clean up a bit. Added option validation with defaults so it is a bit easier to start using the plugin and updated the read me,

There's still one thing left that I would like to fix and that is better connection between the download button and the generated files, I'm thinking maybe we can output a txt with all the url to pfd file path mappings and load that in the frontend?

Cyriuz commented:
Storing the url to file path mapping worked perfectly. I think I have implemented all features that I need now and done with most cleanups, so feel free to feedback!

Cyriuz commented on src/index.ts:
```diff
@@ -160,31 +145,31 @@ export default function (
           }
   
           $(window).on('load', function () {
-            setInterval(checkAndInsertPdfButtons, 1000);
```
Maybe I broke something by removing this interval, but I couldn't figure out why it was running all the time every second?

Cyriuz commented on src/types.ts:
```diff
@@ -5,37 +5,94 @@
  * LICENSE file in the root directory of this source tree.
  */
 
+export interface PluginOptions {
```
I copied this pattern from another plugin I saw. Its not awesome with the duplication but it makes it very neat since it can validate partial configurations and then still be sure in the code that the properties exists.

berndschimmer commented:
any progress here?

Cyriuz commented:
> any progress here?

Please feel free to try it out :) Just clone the branch, run `npm run build` in the folder and then `"docusaurus-plugin-papersaurus": "file:path/to/docusaurus-plugin-papersaurus",` in your docusaurus package.json

abs-lme commented:
Hi @Cyriuz, thanks for your efforts to fix a few bugs 🙏  I was just about to create a MWE for the bug that generating PDFs is not at all possible when I found your fork and PR 😅

I tested your version with my MWE but it still does not work. I get an error from the package `he`:
![image](https://github.com/simologos/docusaurus-plugin-papersaurus/assets/131264015/5bd116c5-5455-481b-8225-da3a3ee050d0)

The function argument `html` is always empty for me.   
It's used here in your code:
![image](https://github.com/simologos/docusaurus-plugin-papersaurus/assets/131264015/676494bb-d13b-44e4-9a7e-875327ab261d)
As this is just about getting the PDF file name maybe it would be OK to use some generic name here?
I don't really know why the HTML is empty for me.

BTW: instead of copying the properties from the `PapersaurusPluginOptions` type [here](https://github.com/Cyriuz/docusaurus-plugin-papersaurus/blob/9825e198c63064486803237ccfbe73e58f44fb6f/src/types.ts#L8C1-L34C2) you can use a [utility type](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype):
```ts
export type PluginOptions = Partial<PapersaurusPluginOptions>
```

Cyriuz commented:
@abs-lme Might it be that you haven't set `projectName: 'Project docs'` in your docusaurus main config? I can fix so that it has a default!

I'm not very used to TypeScript so I'll look into the partial thing, thanks!

abs-lme commented:
@Cyriuz I did set a projectName. I copied a template docusaurus project and added the papersaurus plugin to it. But it won't generate the PDF. The error still persists after pulling.

This is my MWE used for testing:
https://github.com/abs-lme/mwe-docusaurus-papersaurus

Thanks for your efforts!

Cyriuz commented:
@abs-lme Should be fixed now :) I also added some default styles for the toc to look better without any custom styles.

abs-lme commented:
@Cyriuz  thanks a lot! One more issue: the download buttons don't appear any more, and there is an error on the console about `/handbook/pdfs.json` not being loaded.
I assume `handbook` is your test documentation base URL?

I fixed the issue and tested locally. Seems to work now (with and without custom base URL).
Instead of doing another PR roundtrip (me forking your repo and PRing my change) would you mind including the change yourself?

![image](https://github.com/simologos/docusaurus-plugin-papersaurus/assets/131264015/04becfff-c80f-4ba8-96a8-73c19d2fda0c)

179-182:
```ts
          $(window).on('load', function () {
            fetch('${siteConfig.baseUrl}${
            siteConfig.baseUrl?.endsWith("/") ? "" : "/"
          }pdfs.json')
```
I had to add the check for trailing `/` as (contrary to the docusaurus documentation) the `baseUrl` sometimes has no trailing slash.

simologos commented:
@Cyriuz I wanted to thank you very much for the work you are putting into this project!
I initially created Papersaurus for the use in the company where I work, and we still do. One team is currently reviewing your changes. When they don't find regressions, I'm more than happy to merge your changes.

Thank you again :-)

Cyriuz commented:
@abs-lme Yeah sorry I forgot to change that, thanks for fixing! But are you sure baseurl doesn't have a backslash? If that is the case the above code would also fail, maybe I can make a util for it instead.

@simologos For sure, thank you for sharing it in the first place :) I needed it so just wanted to share in case it could be useful for others! There are for sure a few regressions, that's why I made it a draft PR, but if people feel it is worth it I can probably make some backwards compat if needed.

pepek128 commented:
Hey, any progress on that?

Cyriuz commented:
> Please feel free to try it out :) Just clone the branch, run `npm run build` in the folder and then `"docusaurus-plugin-papersaurus": "file:path/to/docusaurus-plugin-papersaurus",` in your docusaurus package.json

@pepek128 from my perspective it is fully functional but it would be good to have more people try it to see if it works :)

berndschimmer commented:
please merge this branch 🙏

simologos commented:
Sorry for the delay here. I reserve time for next week to work on this and merge if all tests are successful.

simologos commented:
Unfortunately my tests were not successful. When calling `getPdfFileName(...)`, the `sideBarItem.unversionedId` is `undefined`.
Fixing this to not throw results in empty PDFs for my docs.

In addition I see a drastic performance decrease when building my documentation.

I'm investing more time to find solutions, but before these two things are fixed, I'm not yet able to merge the PR.

Cyriuz commented:
Thanks for checking! What's your config and docusaurus version? I could probably take a look but its a while ago now so I have to refresh to be able to know what's going on :)

simologos commented:
Thank you Cyriuz. I'm also investigating, but my time is very limited.
I'm currently on docusaurus 2.3.0.

What I see so far is that the `let sidebar = versionInfo.sidebars[sidebarName];` does not contain any more information than `type` and `id`:
```
"sidebars": {
          "mySideBarName": [
            {
              "type": "category",
              "label": "Category Name",
              "items": [
                {
                  "type": "doc",
                  "id": "mySideBarName/id"
                },
              ],
              "collapsed": true,
              "collapsible": true
            },
          ]
        }
```
The `versionInfo` object however, contains a `docs` array, there I see the field `unversionedId` among many others. One thing that may be special is, that we are using multiple sidebars.

Cyriuz commented:
The ids are fetched from the docs here: https://github.com/simologos/docusaurus-plugin-papersaurus/blob/d94055d77f1749d5a80a9dddadc78c30369052ec/src/generate.ts#L241

Could it be that you have a sidebar item with an id that doesn't correspond to an existing doc?

Cyriuz commented:
@simologos I respect that you're low on time so please feel free to just dump your config if you're able to and I can try to get it fixed :)

simologos commented:
Hi @Cyriuz 
Sorry again for the delay with your PR.
I finished my troubleshooting and found that with versioned docs, only the next version successfully created PDFs. For all other versions (current and previous) the article was empty.

With the fix my tests were successful so I'm going to merge your PR.

Thank you very much for your work and sorry for the inconvenience caused.

Cyriuz commented:
Cool, thanks! Interesting find, does that happen in case you have non-auto generated sidebars for a version where the id is specified manually? I wonder if maybe it should be the same in the `category` case above as well but I guess we'll see if anyone gets that problem. I was briefly looking at porting this to work for docusaurus v3 and there the unversionedId didn't even seem exist... Will see if I get around to upgrading :)

simologos commented:
That is possible. We indeed have non-auto generated sidebars only. I hope my change did not break anything for auto-generated sidebars...
Your point is also valid, so far I haven't seen problems with the category, therefore I didn't touch it...

The upgrade to V3 is also on our agenda, but I'm afraid I'm not able to provide a roadmap yet...

Cyriuz commented:
Seems to work fine for me with only auto generated so it's probably correct like that! I'll let you know if I manage to update to v3 as well.

simologos commented:
awesome! thank you very much! once more :-)
